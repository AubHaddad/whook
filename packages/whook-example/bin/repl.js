import { runREPL } from '../src/repl.ts';

await runREPL();
