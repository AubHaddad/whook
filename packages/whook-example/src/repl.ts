import { prepareEnvironment } from './index.js';
import { runREPL as runBaseREPL } from '@whook/whook';

/* Architecture Note #6: REPL

Here is a simple REPL leveraging the depency injection
 features in order to let you test things up easily.
*/
export async function runREPL(): Promise<void> {
  await runBaseREPL(prepareEnvironment);
}
