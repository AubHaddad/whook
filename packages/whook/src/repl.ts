import { printStackTrace } from 'yerror';
import type { Knifecycle, Dependencies } from 'knifecycle';

export async function runREPL<
  D extends Dependencies,
  T extends Knifecycle = Knifecycle,
>(
  prepareEnvironment: ($?: T) => Promise<T>,
  injectedNames: string[] = [],
): Promise<D> {
  try {
    const $ = await prepareEnvironment();
    const services = await $.run([
      ...new Set(['repl', 'log', 'process', ...injectedNames]),
    ]);

    return { $instance: $, ...services } as unknown as D;
  } catch (err) {
    // eslint-disable-next-line
    console.error(
      'ðŸ’€ - Cannot launch the process:',
      printStackTrace(err as Error),
    );
    process.exit(1);
  }
}
