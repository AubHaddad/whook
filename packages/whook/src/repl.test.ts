import { constant } from 'knifecycle';
import { prepareEnvironment as basePrepareEnvironment } from './index';
import { runREPL } from './repl';
import initREPL from './services/repl';
import { PassThrough } from 'stream';

describe('runREPL', () => {
  it('should work', async () => {
    const logger = {
      output: jest.fn(),
      error: jest.fn(),
      debug: jest.fn(),
    };
    const stdin = new PassThrough();
    const stdout = new PassThrough();

    async function prepareEnvironment() {
      const $ = await basePrepareEnvironment();

      $.register(constant('ENV', {}));
      $.register(constant('NODE_ENV', 'test'));
      $.register(constant('DEBUG_NODE_ENVS', []));
      $.register(constant('NODE_ENVS', ['test']));
      $.register(constant('logger', logger));
      $.register(constant('stdin', stdin));
      $.register(constant('stdout', stdout));
      $.register(initREPL);

      return $;
    }
    process.env.ISOLATED_ENV = '1';

    const { $instance } = await runREPL(prepareEnvironment);

    await $instance.destroy();

    expect({
      debugCalls: logger.debug.mock.calls.sort(sortLogs),
      logInfoCalls: logger.output.mock.calls,
      logErrorCalls: logger.error.mock.calls,
    }).toMatchSnapshot();
  });
});

function sortLogs(strs1, strs2) {
  return strs1[0] > strs2[0] ? 1 : strs1[0] === strs2[0] ? 0 : -1;
}
