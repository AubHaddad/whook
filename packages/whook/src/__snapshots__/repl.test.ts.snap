// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runREPL should work 1`] = `
{
  "debugCalls": [
    [
      "ðŸ‘£ - Logging service initialized.",
    ],
    [
      "ðŸ“‡ - Process service initialized.",
    ],
    [
      "ðŸ–µ - Initializing the REPL service!",
    ],
  ],
  "logErrorCalls": [],
  "logInfoCalls": [],
}
`;
